### ISSUE TITLE
Wrong scheduled time for start date in local timezone

### ISSUE THREAD

**Issue by lbatteau** [kubeflow/pipelines#3935]
### What steps did you take:
Created a recurring run with:
* periodic trigger
* run every hour
* start date tomorrow at 00:00 local time (CEST +02:00)

### What happened:
The job was scheduled for tomorrow at 22:00 UTC

### What did you expect to happen:
The job to be scheduled for today at 22:00 UTC

### Environment:
We are running in GKE, standalone KFP
KFP version: : 1d86111
Chrome: 83.0.4103.97 

### Anything else you would like to add:
I believe the same bug is causing another issue. If I schedule a bi-weekly job that should start in a few days at 00:00 CEST, the job is scheduled for two weeks later the next day 22:00 UTC.

/kind bug

**Comment by Bobgy** [comment]
Thanks for the bug report!
Can you give full date of these events in both UTC and local time? so that I won't have a misunderstanding, because I'm not familiar with timezone conversions

**Comment by lbatteau** [comment]
Sure, 

Create a recurring run with:

* periodic trigger
* run every hour
* start date Monday June 22nd at 00:00 CEST, aka Sunday June 21st at 22:00 UTC

What happens:

The job is scheduled for Monday June 22nd at 22:00 UTC

What do you expect to happen:

The job to be scheduled for Sunday 21st 22:00 UTC

**Comment by Bobgy** [comment]
Thanks @lbatteau! That's really helpful.

**Comment by Bobgy** [comment]
Welcome contribution before we can have bandwidth to make a fix.

**Comment by NikeNano** [comment]
I will start to look at this.

**Comment by NikeNano** [comment]
After looking at the code the issue seems to be that time is handled through Unix timestamps in combination with UTC. I assume that we would like to use the local timezone of the cluster or maybe this should actually be specified by the user?

The "current" time used by the controller for deciding if it is time to run a scheduled workflow is collected the following way, code [here](https://github.com/kubeflow/pipelines/blob/7f20a3e843c6c585a7d979f029942838e528f544/backend/src/crd/controller/scheduledworkflow/controller.go#L390): 

```
nowEpoch := c.time.Now().Unix()
```
From the [documentation](https://golang.org/src/time/time.go?s=34918:34944#L1117): ```The result does not depend on the location associated with t```, thus it will not keep any information about timezone. 

This timestamp is compared to the timestamp of the next scheduled run of the pipeline [here](https://github.com/kubeflow/pipelines/blob/7f20a3e843c6c585a7d979f029942838e528f544/backend/src/crd/controller/scheduledworkflow/util/scheduled_workflow.go#L227) which is also an Unix timestamp, which is specifically stated to be UTC(should in reality don't matter, [example](https://play.golang.org/p/QVYwaXtql9X)), link [here](https://github.com/kubeflow/pipelines/blob/7f20a3e843c6c585a7d979f029942838e528f544/backend/src/crd/controller/scheduledworkflow/util/cron_schedule.go#L92).  

As I understand it we have to move away from Unix timestamps in order to keep timezone information. I suggest we instead use variables of the type ```time.time``` and use the local timezone of the cluster to start with. 

What are your thoughts @Bobgy and @jingzhang36?

**Comment by Bobgy** [comment]
I don't see a problem using unix epoch during backend computation, at least for periodic schedule.

Can you elaborate where the conversion went wrong?

But for cron schedule, I agree we need a timezone

**Comment by NikeNano** [comment]
> I don't see a problem using unix epoch during backend computation, at least for periodic schedule.
> 

I was a bit trigger happy and unclear here. I agree with you, we don't need or should swap unix epochs for all the processing. Will clarify the issue below. 

> Can you elaborate where the conversion went wrong?
> 

The reference time which is used for comparing if it is time to run a scheduled workflow is based upon Unix epochs(and thus UTC): 

https://github.com/kubeflow/pipelines/blob/7f20a3e843c6c585a7d979f029942838e528f544/backend/src/crd/controller/scheduledworkflow/controller.go#L390

and the next schedule run is as well

https://github.com/kubeflow/pipelines/blob/7f20a3e843c6c585a7d979f029942838e528f544/backend/src/crd/controller/scheduledworkflow/util/cron_schedule.go#L92

a local timezone will never be considered, all scheduling is handled through UTC. 

When the cron scheduler will check for a job running each day’s  at 07:00 it will do this based upon UTC. I don't think this is a question related to conversion it self it is more related to that Cron expressions are not timezone aware and that since Unix epochs is used, time is considered in UTC.  

Possible solutions that I see: 
 - use timezone aware timestamps, time.time variables.
 - shift all cron expressions to UTC based upon the local timezone.

Let me know if it is still unclear @Bobgy

**Comment by Bobgy** [comment]
Hi @NikeNano, it makes sense using local timezone for cron jobs, but I'd suggest creating a separate issue for it, because the original reported issue here is using periodic schedule, but still getting this problem. So there should be another problem during timezone conversion.

**Comment by NikeNano** [comment]
> Hi @NikeNano, it makes sense using local timezone for cron jobs, but I'd suggest creating a separate issue for it, because the original reported issue here is using periodic schedule, but still getting this problem. So there should be another problem during timezone conversion.

I had created a separate issue for it before this one where created, but it was closed as a duplicate https://github.com/kubeflow/pipelines/issues/2653. I will reopen and update the name to indicate it relates to Cron Expressions. 

Will check the issues with periodic jobs as well.
